### PR TITLE
Convert code-caller to TS

### PR DESCRIPTION
**Pull request by reteps**

### CODE REVIEW

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 1]
```diff
@@ -21,14 +20,17 @@ import { makeAwsClientConfig } from '../aws.js';
...
-const EXITING = Symbol('EXITING');
-const EXITED = Symbol('EXITED');
+import {
+  FunctionMissingError,
+  type CallType,
+  CREATED,
+  IN_CALL,
+  EXITING,
+  EXITED,
+  WAITING,
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-container.ts:30]: Nit, IMO we don't need to put these in `code-caller-shared`. It's a happy accident that both callers use sorta-kinda the same state machine, but really these are implementation details of each separate code caller.

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 2]
```diff
@@ -148,7 +167,7 @@ export class CodeCallerContainer {
     this.coursePath = null;
     this.forbiddenModules = [];
 
-    this._checkState();
+    this._checkState(undefined);
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-container.ts:170]: This is kinda weird. If you change the `_checkState` type to `allowedStates?: CallerState[]`, you don't need to explicitly specify this argument.

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 3]
```diff
@@ -280,10 +288,10 @@ export class CodeCallerContainer {
     this.stdinStream?.write('\n');
 
     this.state = IN_CALL;
-    this._checkState();
+    this._checkState(undefined);
     this.debug('exit call()');
 
-    return deferred.promise;
+    return deferred.promise as any as { result: any; output: string };
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-container.ts:294]: A safer way to do this is to use generics when constructing the deferred promise:

```ts
const deferred = deferredPromise<{ result: any; output: string }>();
```

The little `{ result: any; output: string }` type is used often enough that you might consider making that it's own type (`CodeCallerResult`?).

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 4]
```diff
@@ -470,7 +470,11 @@ export class CodeCallerContainer {
     this.timeoutID = null;
     this._cleanup();
     this.state = EXITING;
-    this._callCallback(new Error('timeout exceeded, killing CodeCallerContainer container'));
+    this._callCallback(
+      new Error('timeout exceeded, killing CodeCallerContainer container'),
+      undefined,
+      '',
+    );
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-container.ts:477]: As with `_checkState`, it'd be better to make the last two arguments optional.

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 5]
```diff
@@ -111,15 +113,32 @@ async function ensureImage() {
   }
 }
 
-/** @typedef {import('./code-caller-shared.js').CodeCaller} CodeCaller */
-/** @typedef {import('./code-caller-shared.js').CallType} CallType */
-
-/**
- * @implements {CodeCaller}
- */
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-container.ts:119]: Let's not drop the `implements CodeCaller`.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 6]
```diff
@@ -11,45 +11,47 @@ import { logger } from '@prairielearn/logger';
...
+  IN_CALL,
+  RESTARTING,
+  EXITING,
+  EXITED,
+  type CodeCaller,
+  type PrepareForCourseOptions,
+  type CallType as CallTypeShared,
+} from './code-caller-shared.js';
+
+/* Re-export types until executor is in TypeScript are merged */
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:29]: Hopefully we can land the executor-in-TS PR first!

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 7]
```diff
@@ -132,7 +135,7 @@ export class CodeCallerNative {
     this.coursePath = null;
     this.forbiddenModules = [];
 
-    this._checkState();
+    this._checkState(undefined);
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:138]: Same comment as in other file.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 8]
```diff
@@ -237,19 +234,17 @@ export class CodeCallerNative {
     this.child?.stdin?.write('\n');
 
     this.state = IN_CALL;
-    this._checkState();
+    this._checkState(undefined);
     this.debug('exit call()');
 
-    return deferred.promise;
+    return deferred.promise as Promise<{ result: any; output: string }>;
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:240]: Same comment as in other file.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 9]
```diff
@@ -342,16 +337,17 @@ export class CodeCallerNative {
       env.DROP_PRIVILEGES = '1';
     }
 
-    /** @type {import('child_process').SpawnOptions} */
-    const options = {
+    const options: SpawnOptions = {
       cwd: import.meta.dirname,
       // stdin, stdout, stderr, data, and restart confirmations
-      stdio: ['pipe', 'pipe', 'pipe', 'pipe', 'pipe'],
+      stdio: ['pipe', 'pipe', 'pipe', 'pipe', 'pipe'] as IOType[],
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:343]: Is `as IOType[]` needed here? I would have expected the `SpawnOptions` type to automatically allow what we have here.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 10]
```diff
@@ -342,16 +337,17 @@ export class CodeCallerNative {
...
       env,
     };
-    const child = /** @type {CodeCallerNativeChildProcess} */ (
-      child_process.spawn(cmd, args, options)
-    );
+    const child: CodeCallerNativeChildProcess = child_process.spawn(
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:346]: The `as ...` should be sufficient.

```suggestion
    const child = child_process.spawn(
```

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 11]
```diff
@@ -11,45 +11,47 @@ import { logger } from '@prairielearn/logger';
 import { deferredPromise } from '../deferred.js';
 import { APP_ROOT_PATH, REPOSITORY_ROOT_PATH } from '../paths.js';
 
-import { FunctionMissingError } from './code-caller-shared.js';
+import { type CodeCallerNativeChildProcess } from './code-caller-native-types.js';
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:14]: Nit: now that `code-caller-native` is itself in TypeScript, consider inlining this `CodeCallerNativeChildProcess` type in this file.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts [hunk 12]
```diff
@@ -450,6 +446,8 @@ export class CodeCallerNative {
             ', signal = ' +
             String(signal),
         ),
+        undefined,
+        '',
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/code-caller-native.ts:450]: Same comment as in other file.

File: apps/prairielearn/src/lib/code-caller/index.ts [hunk 13]
```diff
@@ -70,7 +70,7 @@ export async function init() {
 
         await codeCaller.ensureChild();
         load.startJob('python_worker_idle', codeCaller.uuid);
-        return codeCaller;
+        return codeCaller as CodeCaller;
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/index.ts:73]: Type inference should make this unnecessary.

```suggestion
        return codeCaller;
```

File: apps/prairielearn/src/lib/code-caller/index.ts [hunk 14]
```diff
@@ -53,12 +52,13 @@ export async function init() {
...
       create: async () => {
         const codeCallerOptions = {
+          dropPrivileges: false,
           questionTimeoutMilliseconds: config.questionTimeoutMilliseconds,
           pingTimeoutMilliseconds: config.workerPingTimeoutMilliseconds,
           errorLogger: logger.error.bind(logger),
         };
 
-        const codeCaller = run(() => {
+        const codeCaller: CodeCallerContainer | CodeCallerNative = run(() => {
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/code-caller/index.ts:61]: Type inference should make this unnecessary. If you really want to be explicit, you can use the `CodeCaller` type instead.

```suggestion
        const codeCaller = run(() => {
```

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts (hunk 3 above)
Comment (reteps) [reply apps/prairielearn/src/lib/code-caller/code-caller-container.ts:302]: Sortof fixed it? had to do `deferred.resolve({ result, output: output ?? '' });` a little higher up.

File: apps/prairielearn/src/lib/code-caller/code-caller-native.ts (hunk 6 above)
Comment (reteps) [reply apps/prairielearn/src/lib/code-caller/code-caller-native.ts:43]: That is in #11011, which just got merged in.

File: apps/prairielearn/src/lib/code-caller/code-caller-container.ts [hunk 15]
```diff
@@ -439,14 +454,11 @@ export class CodeCallerContainer {
     this.container
       .wait()
       .then((status) => this._handleContainerExit(null, status))
-      .catch((err) => this._handleContainerExit(err));
+      .catch((err) => this._handleContainerExit(err, 999));
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/lib/code-caller/code-caller-container.ts:457]: This was a poor decision, i don't know why i added that. 😬

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/lib/code-caller/code-caller-container.ts:457]: Reverted back (good catch 😅 )

**Pull request merged**